asset/asset.go

```diff
@@ -938,6 +957,193 @@ type GroupKeyReveal interface {
 	GroupPubKey(assetID ID) (*btcec.PublicKey, error)
 }
 
+// GroupKeyRevealTapscript represents the data structure used to derive the
+// tweaked tapscript root, which is subsequently used to compute the asset
+// group key.
+//
+// The tapscript tree ensures that the derived asset group key is unique
+// to a specific genesis asset ID. This design prevents asset group keys from
+// being reused across different genesis assets or non-compliant asset minting
+// tranches (e.g., tranches of a different asset type).
+//
+// The uniqueness of the asset group key to a specific genesis asset ID is
+// guaranteed by including the genesis asset ID as a leaf in the tapscript tree.
+// Additionally, we ensure that the tapscript tree cannot include multiple
+// recognizable genesis asset IDs. To achieve this, we enforce that the genesis
+// asset ID must appear in the first leaf layer of the tapscript tree. At this
+// tree level, there are exactly two leaf positions. We use the remaining
+// revealed structure of the tree to prove that the genesis asset ID is
+// correctly committed and that no conflicting or duplicate genesis asset IDs
+// are committed.
+//
+// The tapscript tree can include a custom tapscript subtree, enabling users
+// to incorporate custom script spend paths. However, in the simplest case,
+// if a custom tapscript tree is not included, the tapscript tree adopts the
+// following structure:
+//
+//	           [tapscript root]
```

Comment:
I think there might be some confusion on how the Taproot key creation works today.

This is my sketch of how things work with just Taproot:
```
              (taproot output key)                                 
                       |                                           
       +---------------+----------+                                
       |                          |                                
 (internal key)            [tapscript root]                        
                                  |                                
                     +------------+-------------+                  
                     |                          |                  
               [leaf 1 hash]             [leaf 2 hash]             
                   <leaf 1>                  <leaf 2> 

[] = hash
() = pubkey
<> = script

Where (taproot output key) = (internal key) + TapTweak((internal key) || [tapscript root])*G
```

But with the asset ID as a top-level tree element, it would look like this:

```
             (taproot output key)                                  
                      |                                            
      +---------------+----------+                                 
      |                          |                                 
(internal key)            [tapscript root]                         
                                 |                                 
                    +------------+-------------+                   
                    |                          |                   
              [gen asset ID]                                       
                                        [tapscript sibling hash]   
                                                                   
                                  +------------+-------------+     
                                  |                          |     
                            [leaf 1 hash]             [leaf 2 hash]
                                <leaf 1>                  <leaf 2> 
```

So not sure if the rest of the text below still works with this corrected view?

Reply:
With either diagram, I think we should make it clear where the tapscript merkle tree ends, and the EC tweaking operations start. At a glance for both, it looks like the internal key is actually part of the merkle tree.